src/dataflow-types/src/client/replicated.rs

```diff
@@ -221,8 +224,57 @@ where
+                                    lower: _,
+                                    upper,
+                                    mut updates,
+                                }) => {
+                                    // It is sufficient to compare `upper` against the last reported frontier for `id`,
+                                    // and if `upper` is not less or equal to that frontier, some progress has happened.
+                                    // If so, we retain only the updates greater or equal to that last reported frontier,
+                                    // and announce a batch from that frontier to its join with `upper`.
+
+                                    // Ensure that we have a recorded frontier ready to go.
+                                    let entry = self
+                                        .tails
+                                        .entry(id)
+                                        .or_insert_with(|| Antichain::from_elem(T::minimum()));
+                                    // If the upper frontier has changed, we have a statement to make.
+                                    // This happens if there is any element of `entry` not greater or
+                                    // equal to some element of `upper`.
+                                    use differential_dataflow::lattice::Lattice;
+                                    let new_upper = entry.join(&upper);
+                                    if &new_upper != entry {
+                                        let new_lower = entry.clone();
+                                        entry.clone_from(&new_upper);
+                                        updates.retain(|(time, _data, _diff)| {
+                                            new_lower.less_equal(time)
+                                        });
+                                        return Ok(Some(ComputeResponse::TailResponse(
+                                            id,
+                                            TailResponse::Batch(TailBatch {
+                                                lower: new_lower,
+                                                upper: new_upper,
+                                                updates,
+                                            }),
+                                        )));
+                                    }
+                                }
+                                TailResponse::Dropped => {
+                                    // Introduce a new terminal frontier to suppress all future responses.
+                                    // We cannot simply remove the entry, as we currently create new entries in response
+                                    // to observed responses; if we pre-load the entries in response to commands we can
+                                    // clean up the state here.
+                                    self.tails.insert(id, Antichain::new());
```

Comment:
We eventually need a cleanup mechanism to avoid keeping around state forever, but I think this we can tackle independently of this PR.

Reply:
I agree! This doesn't change the behavior from e.g. the `partitioned` client, which does the same thing. If we are able to sniff out the tail installation commands, we could swing this around, but .. that's all a bit mysterious at the moment.